### DISCUSSION TITLE
Q:how to run a simple photoacoustic image reconstruction example using time reversal?

### DISCUSSION THREAD

**Discussion by vanlinus3068** [waltsims/k-wave-python#528] [Q&A]
hi! This is my  first time starting a discussion in this repository! I want to use the kspaceFirstOrder2D function to simulate the time-reversal algorithm, but I’m not sure how to use or define sensor.time_reversal_boundary_data. Could someone please help me with this?

Following is my code:

import cv2
import numpy as np
from scipy.io import savemat
from kwave.data import Vector
from kwave.kgrid import kWaveGrid
from kwave.kmedium import kWaveMedium
from kwave.ksensor import kSensor
from kwave.ksource import kSource
from kwave.kspaceFirstOrder2D import kspaceFirstOrder2D
from kwave.options.simulation_execution_options import SimulationExecutionOptions
from kwave.options.simulation_options import SimulationOptions
from kwave.utils.mapgen import make_cart_circle
import matplotlib.pyplot as plt

if __name__ == '__main__':

    Nx = 128
    Ny = 128
    dx = 0.1e-3
    dy = 0.1e-3
    kgrid = kWaveGrid(N=(Nx, Ny), spacing=(dx, dy))

    medium = kWaveMedium(sound_speed=1500,
                         density=1000,
                         alpha_coeff=0.75,
                         alpha_power=1.5, )

    img_matrix = cv2.imread('EXAMPLE.png', cv2.IMREAD_GRAYSCALE)

    source = kSource()
    source.p0 = (1 * img_matrix) / 255

    plt.imshow(img_matrix, cmap='gray')
    plt.title('Loaded Image')
    plt.axis('off')
    plt.show()

    sensor_radius: float = 4e-3
    num_sensor_points: int = 50
    sensor_mask = make_cart_circle(radius=sensor_radius, num_points=num_sensor_points, plot_circle=True)
    sensor = kSensor(mask=sensor_mask)
    kgrid.makeTime(c=medium.sound_speed)

    opts = {
        'smooth_p0': True,
        'pml_inside': False
    }

    simul_opts = SimulationOptions(save_to_disk=True, **opts)
    exec_opts = SimulationExecutionOptions(is_gpu_simulation=True)

    result = kspaceFirstOrder2D(kgrid=kgrid, source=source, sensor=sensor, medium=medium,
                                simulation_options=simul_opts, execution_options=exec_opts)

    print('keys in result:', result.keys())
    print('sensor data shape:', result['p'].shape)
    print('dt =', result['dt'])

    sensor_data = result['p']
    np.savez('my_sensor_data.npz', result=sensor_data)
    savemat('my_sensor_data.mat', {'sensor_data': sensor_data})

    plt.imshow(sensor_data.T, cmap='viridis', aspect='auto')
    plt.colorbar()
    plt.title("Sensor Data Visualization")
    plt.xlabel('time idx')
    plt.ylabel('sensor idx')
    plt.show()

    sensor = kSensor(mask=sensor_mask)
    sensor.time_reversal_boundary_data = sensor_data
    sensor.p_mask = sensor_mask
    source.p0 = np.zeros([128, 128])
    source.p = np.zeros([128, 128])

    input_args = {'SaveToDisk': False}
    reconstructed_data = kspaceFirstOrder2D(kgrid, source, sensor, medium, simulation_options=simul_opts,
                                            execution_options=exec_opts)
    initial_pressure = reconstructed_data['p']

**Comment by carmsignes** [comment]
Hey! I am trying to do a similar thing, did you manage to find a way?

**Reply by vanlinus3068** [reply]
Not yet, but I'm here to help if you need it! What are you working on? 😊

**Reply by carmsignes** [reply]
Thanks for your answer! You can check the discussion I opened recently for more details! Basically, I am also using a circular array to detect changes in water temperature. Ideally I want to create a 2d map of the medium, and I also thought of using time reversal image reconstruction. I am building on the array_as_sensor example.  For me, this produces an output I can plot:

sensor.time_reversal_boundary_data = sensor_data

p0_recon = kspaceFirstOrder2D(kgrid, source, sensor, medium, simulation_options, execution_options)
p0_recon = p0_recon['p_final'].T

Regarding the examples, only this has two problems: first the inverse crime, which should be straightforward to fix, and secondly I did not reinitialize the sensor and source, because when I tried, it raised issues. Besides, the output I can plot has the appearance
 of an interference pattern, and I am not sure what exactly is it I am seeing:
 
![interference pattern](https://github.com/user-attachments/assets/451a4662-a37a-4ff8-9b35-9bf12a97c04e)

What was your problem with the code you provided?

**Reply by vanlinus3068** [reply]
I want to run a photoacoustic imaging time-reversal algorithm with a circular sensor array, similar to the example in the k-Wave MATLAB Toolbox (http://www.k-wave.org/documentation/example_pr_2D_tr_circular_sensor.php). However, it's running slowly, and I'm not sure if it's correct. Additionally, I don't know how to plot the images reconstructed by TR .  In addition, this issue has been raised before and cannot be solved. Therefore, I found another method, DAS algorithm, to reconstruct images.It is simple and easy to use. Do u want to try?

**Reply by vanlinus3068** [reply]
It was great meeting you. I'd love to keep in touch. May I have your contact information, like your social media account or any other instant messaging app you use？

**Reply by carmsignes** [reply]
Are you using a phantom for the medium? For me the calculations do not take much time, but I am using a rather simple medium. 
As for the TR images, the picture I showed was produced through, although I am not exactly sure if it is the way to do it:
```
p0_recon = kspaceFirstOrder2D(kgrid, source, sensor, medium, simulation_options, execution_options)
p0_recon = p0_recon['p_final'].T
fig, ax = plt.subplots(1, 1)
im = ax.imshow(p0_recon,
           extent=[kgrid.y_vec.min() * 1e3, kgrid.y_vec.max() * 1e3, kgrid.x_vec.max() * 1e3, kgrid.x_vec.min() * 1e3],
           vmin=-1e-8, vmax=1e-8, cmap=cmap)
divider = make_axes_locatable(ax)
cax = divider.append_axes("right", size="3%", pad="2%")
ax.set_ylabel('x-position [mm]')
ax.set_xlabel('y-position [mm]')
fig.colorbar(im, cax=cax)
plt.show()
```
Could you point out where the issue was raised before and not solved? And yes, definitely! Could you send me more info on how to implement the  DAS algorithm?

**Reply by vanlinus3068** [reply]
my code：
`from copy import deepcopy

import cv2
import matplotlib.pyplot as plt
from kwave.kgrid import kWaveGrid
from kwave.kmedium import kWaveMedium
from kwave.ksensor import kSensor
from kwave.ksource import kSource
from kwave.kspaceFirstOrder2D import kspace_first_order_2d_gpu, kspaceFirstOrder2D
from kwave.options.simulation_execution_options import SimulationExecutionOptions
from kwave.options.simulation_options import SimulationOptions
from kwave.utils.mapgen import make_cart_circle

if __name__ == '__main__':
    Nx = 128
    Ny = 128
    dx = 0.1e-3
    dy = 0.1e-3
    kgrid = kWaveGrid(N=(Nx, Ny), spacing=(dx, dy))


    medium = kWaveMedium(sound_speed=1500,
                         density=1000,
                         alpha_coeff=0.75,
                         alpha_power=1.5, )

    img_matrix = cv2.imread('EXAMPLE.png', cv2.IMREAD_GRAYSCALE)

    source = kSource()
    source.p0 = (1 * img_matrix) / 255

    plt.imshow(img_matrix, cmap='gray')
    plt.title('Loaded Image')
    plt.axis('off')
    plt.show()

    sensor_radius: float = 4e-3
    num_sensor_points: int = 50
    sensor_mask = make_cart_circle(radius=sensor_radius, num_points=num_sensor_points, plot_circle=True)
    sensor = kSensor(mask=sensor_mask)
    kgrid.makeTime(c=medium.sound_speed)

    opts = {
        'smooth_p0': True,
        'pml_inside': False
    }

    simul_opts = SimulationOptions(save_to_disk=True, **opts)
    exec_opts = SimulationExecutionOptions(is_gpu_simulation=True)
    result = kspace_first_order_2d_gpu(kgrid=kgrid, source=source, sensor=sensor, medium=medium,
                                simulation_options=simul_opts, execution_options=exec_opts)
    sensor_data = result['p']
    
    plt.imshow(sensor_data.T, cmap='viridis', aspect='auto')
    plt.colorbar()
    plt.title("Sensor Data Visualization")
    plt.xlabel('time idx')
    plt.ylabel('sensor idx')
    plt.show()
    
    # run the time-reversal reconstruction
    sensor = kSensor()
    sensor.mask = sensor_mask
    sensor.p_mask = sensor_mask
    sensor.time_reversal_boundary_data = sensor_data
    p0_recon_interp = kspaceFirstOrder2D(medium=deepcopy(medium),
                                         kgrid=deepcopy(kgrid),
                                         source=deepcopy(source),
                                         sensor=deepcopy(sensor),
                                         simulation_options=deepcopy(simul_opts),
                                         execution_options=deepcopy(exec_opts))



`
Then it raise issue：
`Traceback (most recent call last):
  File "D:\PythonProject\20241209-examples\test3-simple.py", line 66, in <module>
    p0_recon_interp = kspaceFirstOrder2D(medium=deepcopy(medium),
  File "D:\miniconda\envs\pytorch39\lib\site-packages\kwave\kspaceFirstOrder2D.py", line 295, in kspaceFirstOrder2D
    k_sim = kWaveSimulation(kgrid=kgrid, source=source, sensor=sensor, medium=medium, simulation_options=simulation_options)
  File "D:\miniconda\envs\pytorch39\lib\site-packages\kwave\kWaveSimulation.py", line 53, in __init__
    source = {"p_mask": self.sensor.p_mask, "p": np.flip(self.sensor.time_reversal_boundary_data, 2), "p_mode": "dirichlet"}
  File "D:\miniconda\envs\pytorch39\lib\site-packages\numpy\lib\function_base.py", line 340, in flip
    axis = _nx.normalize_axis_tuple(axis, m.ndim)
  File "D:\miniconda\envs\pytorch39\lib\site-packages\numpy\core\numeric.py", line 1380, in normalize_axis_tuple
    axis = tuple([normalize_axis_index(ax, ndim, argname) for ax in axis])
  File "D:\miniconda\envs\pytorch39\lib\site-packages\numpy\core\numeric.py", line 1380, in <listcomp>
    axis = tuple([normalize_axis_index(ax, ndim, argname) for ax in axis])
numpy.exceptions.AxisError: axis 2 is out of bounds for array of dimension 2`


My DAS：
`def reconstruction_das(sensor_data, radius=4e-3, image_shape=(128, 128), dx=0.1e-3, img_show=False):

    X, Y = image_shape
    dx = dx
    dy = dx
    fs = 50e6
    dt = 1 / fs
    N = sensor_data.shape[0]
    angle = 2 * np.pi / (N - 1)
    R = radius
    vs = 1.5e3

    reconstructed_IM = np.zeros((X, Y))
    for i in range(X):
        for j in range(Y):
            for channel_i in range(N):
                d = np.sqrt(((i - X / 2) * dx - R * np.cos(channel_i * angle - np.pi)) ** 2 +
                            ((j - Y / 2) * dy - R * np.sin(channel_i * angle - np.pi)) ** 2)
                time = int(np.floor(d / vs * fs)) + 1
                if time < sensor_data.shape[1]:
                    reconstructed_IM[i, j] += sensor_data[channel_i, time]


    if img_show:
        plt.figure()
        plt.imshow(sensor_data, aspect='auto', cmap='jet', extent=[0, sensor_data.shape[1], 0, N])
        plt.title('Sensor Signals Over Time')
        plt.xlabel('Time Step')
        plt.ylabel('Sensor Index')
        plt.colorbar()
        plt.show()

    if img_show:
        plt.figure()
        plt.imshow(reconstructed_IM, extent=[0, X * dx * 1e3, 0, Y * dy * 1e3])
        plt.title('Reconstructed PAT image')
        plt.ylabel('x-position [mm]')
        plt.xlabel('y-position [mm]')
        plt.axis('image')
        plt.colorbar()
        plt.show()
    return reconstructed_IM`
Hopefully these codes will help！
